From dd794489f97baf760d03b32e4e3188b5af799436 Mon Sep 17 00:00:00 2001
From: Michal Sekletar <msekleta@redhat.com>
Date: Wed, 7 Sep 2022 17:37:34 +0200
Subject: [PATCH] core: add possibility to not track certain unit types

(cherry picked from commit 88e4bfa62bd2561e04a90dc009e7a3865e0878fb)

Related: RHEL-5877
---
 src/core/unit.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/core/unit.c b/src/core/unit.c
index d3459dcdd0..ac960ef0c8 100644
--- a/src/core/unit.c
+++ b/src/core/unit.c
@@ -18,6 +18,7 @@
 #include "dbus-unit.h"
 #include "dbus.h"
 #include "dropin.h"
+#include "env-util.h"
 #include "escape.h"
 #include "execute.h"
 #include "fd-util.h"
@@ -4786,11 +4787,28 @@ int unit_setup_dynamic_creds(Unit *u) {
 }
 
 bool unit_type_supported(UnitType t) {
+        static int8_t cache[_UNIT_TYPE_MAX] = {}; /* -1: disabled, 1: enabled: 0: don't know */
+        int r;
+
         if (_unlikely_(t < 0))
                 return false;
         if (_unlikely_(t >= _UNIT_TYPE_MAX))
                 return false;
 
+        if (cache[t] == 0) {
+                char *e;
+
+                e = strjoina("SYSTEMD_SUPPORT_", unit_type_to_string(t));
+
+                r = getenv_bool(ascii_strupper(e));
+                if (r < 0 && r != -ENXIO)
+                        log_debug_errno(r, "Failed to parse $%s, ignoring: %m", e);
+
+                cache[t] = r == 0 ? -1 : 1;
+        }
+        if (cache[t] < 0)
+                return false;
+
         if (!unit_vtable[t]->supported)
                 return true;
 
